src/main/java/jcifs/spnego/NegTokenInit.java

```diff
@@ -0,0 +1,226 @@
+        parse(token);
+    }
+
+
+    public int getContextFlags () {
+        return this.contextFlags;
+    }
+
+
+    public void setContextFlags ( int contextFlags ) {
+        this.contextFlags = contextFlags;
+    }
+
+
+    public boolean getContextFlag ( int flag ) {
+        return ( getContextFlags() & flag ) == flag;
+    }
+
+
+    public void setContextFlag ( int flag, boolean value ) {
+        setContextFlags(value ? ( getContextFlags() | flag ) : ( getContextFlags() & ( 0xffffffff ^ flag ) ));
+    }
+
+
+    public ASN1ObjectIdentifier[] getMechanisms () {
+        return this.mechanisms;
+    }
+
+
+    public void setMechanisms ( ASN1ObjectIdentifier[] mechanisms ) {
+        this.mechanisms = mechanisms;
+    }
+
+
+    /**
+     * {@inheritDoc}
+     *
+     * @see java.lang.Object#toString()
+     */
+    @Override
+    public String toString () {
+        String mic = null;
+        if ( this.getMechanismListMIC() != null ) {
+            mic = Hexdump.toHexString(this.getMechanismListMIC(), 0, this.getMechanismListMIC().length);
+        }
+        return String.format("NegTokenInit[flags=%d,mechs=%s,mic=%s]", this.getContextFlags(), Arrays.toString(this.getMechanisms()), mic);
+    }
+
+
+    @Override
+    public byte[] toByteArray () {
+        try {
+            ASN1EncodableVector fields = new ASN1EncodableVector();
+            ASN1ObjectIdentifier[] mechs = getMechanisms();
+            if ( mechs != null ) {
+                ASN1EncodableVector vector = new ASN1EncodableVector();
+                for ( int i = 0; i < mechs.length; i++ ) {
+                    vector.add(mechs[ i ]);
+                }
+                fields.add(new DERTaggedObject(true, 0, new DERSequence(vector)));
+            }
+            int ctxFlags = getContextFlags();
+            if ( ctxFlags != 0 ) {
+                fields.add(new DERTaggedObject(true, 1, new DERBitString(ctxFlags)));
+            }
+            byte[] mechanismToken = getMechanismToken();
+            if ( mechanismToken != null ) {
+                fields.add(new DERTaggedObject(true, 2, new DEROctetString(mechanismToken)));
+            }
+            byte[] mechanismListMIC = getMechanismListMIC();
+            if ( mechanismListMIC != null ) {
+                fields.add(new DERTaggedObject(true, 3, new DEROctetString(mechanismListMIC)));
+            }
+
+            ASN1EncodableVector ev = new ASN1EncodableVector();
+            ev.add(SPNEGO_OID);
+            ev.add(new DERTaggedObject(true, 0, new DERSequence(fields)));
+            ByteArrayOutputStream collector = new ByteArrayOutputStream();
+            DEROutputStream der = new DEROutputStream(collector);
+            DERApplicationSpecific derApplicationSpecific = new DERApplicationSpecific(0, ev);
+            der.writeObject(derApplicationSpecific);
+            return collector.toByteArray();
+        }
+        catch ( IOException ex ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)